engines/director/lingo/xlibs/msfile.cpp

```diff
@@ -0,0 +1,337 @@
+		volLabel = "Simpsons";
+	} else {
+		warning("MSFile::m_getVolLabel(): Unsupported gameid '%s'", gameId);
+	}
+
+	g_lingo->push(volLabel);
+}
+
+XOBJSTUB(MSFile::m_createDir, 0)
+XOBJSTUB(MSFile::m_removeDir, 0)
+XOBJSTUB(MSFile::m_getCurrDir, "")
+XOBJSTUB(MSFile::m_setCurrDir, 0)
+XOBJSTUB(MSFile::m_removeFile, 0)
+XOBJSTUB(MSFile::m_renameFile, 0)
+
+Common::Path MSFile::resolveSourceFilePath(const Common::String& srcParam) {
+	return findPath(srcParam);
+}
+
+Common::SeekableReadStream *MSFile::openSourceStream(const Common::Path &srcFilePath, const Common::String &srcParam, char dirSeparator) {
+	Common::SaveFileManager *saves = g_system->getSavefileManager();
+	Common::String prefix = savePrefix();
+
+	// The source file may be a save file in the ScummVM save folder.
+	Common::String saveFileName = lastPathComponent(srcParam, dirSeparator);
+
+	// Modify the source file name to match the name of the potential save file
+	if (!saveFileName.hasSuffixIgnoreCase(".txt")) {
+		saveFileName += ".txt";
+	}
+
+	if (!saveFileName.hasPrefixIgnoreCase(prefix)) {
+		saveFileName = prefix + saveFileName;
+	}
+
+	// Try to open the save file
+	Common::SeekableReadStream *srcReadStream = saves->openForLoading(saveFileName);
+
+	// If the save file was not found then we will try opening the file directly
+	if (!srcReadStream) {
+		// We have a relative path so we can open it using Common::File
+		Common::File *srcFile = new Common::File();
+		if (srcFilePath.empty() || !srcFile->open(srcFilePath)) {
+			warning("MSFile::m_copyFile(): Error opening file %s", srcFilePath.toString(Common::Path::kNativeSeparator).c_str());
+			delete srcFile;
+			return nullptr;
+		}
+		srcReadStream = srcFile;
+	}
+
+	return srcReadStream;
+}
+
+Common::SeekableWriteStream *MSFile::openDestinationStream(Common::String &destFileName) {
+	Common::SaveFileManager *saves = g_system->getSavefileManager();
+	Common::String prefix = savePrefix();
+
+	Common::OutSaveFile *destFile = nullptr;
+
+	if (!destFileName.hasSuffixIgnoreCase(".txt")) {
+		destFileName += ".txt";
+	}
+
+	if (!destFileName.hasPrefixIgnoreCase(prefix)) {
+		destFileName = prefix + destFileName;
+	}
+
+	destFile = saves->openForSaving(destFileName, false);
+
+	return destFile;
+}
+
+bool MSFile::copyStream(Common::SeekableReadStream* srcStream, Common::SeekableWriteStream* destStream) {
+	uint32 size = 0;
+	void *buffer = nullptr;
+
+	if (srcStream && destStream) {
+		size = srcStream->size();
+
+		// Allocate a buffer to hold the read file contents
+		buffer = malloc(size);
+		assert(buffer);
+
+		// Read the full file contents into the buffer
+		srcStream->read(buffer, size);
+		bool error = srcStream->err();
+		delete srcStream;
+
+		if (error) {
+			free(buffer);
+			return false;
+		}
+
+		// Write the source file buffer to the destination file. This is effectively a file copy.
+		destStream->write(buffer, size);
```

Comment:
You could use `destStream->writeStream(srcStream);` to simplify the code and avoid creating larger buffers in memory.